components/abstractions/src/main/java/com/microsoft/kiota/store/InMemoryBackingStore.java

```diff
@@ -212,39 +234,61 @@ private void setupNestedSubscriptions(
         }
     }
 
-    private void ensureCollectionPropertyIsConsistent(final String key, final Object storeItem) {
-        if (storeItem instanceof Pair) { // check if we put in a collection annotated with the size
-            final Pair<?, Integer> collectionTuple = (Pair<?, Integer>) storeItem;
-            Object[] items;
-            if (collectionTuple.getValue0() instanceof Collection) {
-                items = ((Collection<Object>) collectionTuple.getValue0()).toArray();
-            } else { // it is a map
-                items = ((Map<?, Object>) collectionTuple.getValue0()).values().toArray();
-            }
+    private void ensureCollectionPropertiesAreConsistent() {
+        final HashMap<String, Object> currentStoreDirtyCollections = new HashMap<>();
+        final List<BackedModel> nestedBackedModelsToEnumerate = new ArrayList<>();
 
-            for (final Object item : items) {
-                touchNestedProperties(item); // call get on nested properties
+        for (final Map.Entry<String, Pair<Boolean, Object>> entry : this.store.entrySet()) {
+            final Pair<Boolean, Object> wrapper = entry.getValue();
+            if (isCollectionValue(wrapper)) {
+                final Pair<?, Integer> collectionTuple = (Pair<?, Integer>) wrapper.getValue1();
+                Object[] items = getObjectArrayFromCollectionWrapper(collectionTuple);
+                for (final Object item : items) {
+                    if (!(item instanceof BackedModel)) break;
+                    nestedBackedModelsToEnumerate.add((BackedModel) item);
```

Comment:
Similar concern here, the cast will most likely fail if we called `getObjectArrayFromCollectionWrapper` on a map that has mixed types and the first item happened to be a backed model.

Reply:
Thanks for catching this. Made these updates & added tests.

Tests also exposed a bug where we now need to propagate `returnOnlyChangedValues` to nested `BackedModels` to ensure nested models with collection properties are checked. We have limited the collection checks to happen during enumeration only when we want to return changed values.